src/lib.rs

```diff
@@ -14,46 +14,110 @@
+    Identity,
+    /// No preference is expressed on which encoding to use. Either the `Accept-Encoding` header is not present, or `*` is set as the most preferred encoding.
+    None,
+}
+
+impl Encoding {
+    /// Parses a given string into its corresponding encoding.
+    fn parse(s: &str) -> Result<Encoding> {
+        match s {
+            "gzip" => Ok(Encoding::Gzip),
+            "deflate" => Ok(Encoding::Deflate),
+            "br" => Ok(Encoding::Brotli),
+            "identity" => Ok(Encoding::Identity),
+            "*" => Ok(Encoding::None),
+            _ => Err(ErrorKind::UnknownEncoding)?,
+        }
+    }
+
+    /// Converts the encoding into its' corresponding header value.
+    ///
+    /// Note that [`Encoding::None`] will return a HeaderValue with the content `*`.
+    /// This is likely not what you want if you are using this to generate the `Content-Encoding` header to be included in an encoded response.
+    pub fn to_header_value(&self) -> HeaderValue {
+        match *self {
+            Encoding::Gzip => HeaderValue::from_str("gzip").unwrap(),
+            Encoding::Deflate => HeaderValue::from_str("deflate").unwrap(),
+            Encoding::Brotli => HeaderValue::from_str("br").unwrap(),
+            Encoding::Identity => HeaderValue::from_str("identity").unwrap(),
+            Encoding::None => HeaderValue::from_str("*").unwrap(),
+        }
+    }
 }
 
 /// Parse a set of HTTP headers into an `Encoding`.
 pub fn parse(headers: &HeaderMap) -> Result<Encoding> {
-  let header = match headers.get(ACCEPT_ENCODING) {
-    Some(header) => header,
-    None => return Ok(Encoding::None),
-  };
-
-  let string = header.to_str().context(ErrorKind::InvalidEncoding)?;
-
-  if string.contains("br") {
-    Ok(Encoding::Brotli)
-  } else if string.contains("deflate") {
-    Ok(Encoding::Deflate)
-  } else if string.contains("gzip") {
-    Ok(Encoding::Gzip)
-  } else {
-    Err(ErrorKind::UnknownEncoding)?
-  }
+    let mut preferred_encoding = Encoding::None;
+    let mut max_qval = 0.0;
+
+    for header_value in headers.get_all(ACCEPT_ENCODING).iter() {
+        let header_value = header_value.to_str().context(ErrorKind::InvalidEncoding)?;
+        for v in header_value.split(',').map(str::trim) {
+            let v: Vec<&str> = v.splitn(2, ";q=").collect();
+            let encoding = v[0];
+
+            match Encoding::parse(encoding) {
+                Ok(encoding) => {
+                    if v.len() > 1 {
+                        let qval = match v[1].parse::<f32>() {
+                            Ok(f) => f,
+                            Err(_) => continue, // skip malformed q values
+                        };
+                        if (qval - 1.0f32).abs() < 0.01 {
+                            preferred_encoding = encoding;
+                            break;
+                        } else if qval > 1.0 {
+                            // q-values over 1 are unacceptable
+                            continue;
```

Comment:
Should this error instead? It seems like malformed user input to me.

Reply:
It probably should, on second thought -- I'll rewrite this